### PR TITLE
Add support for execution spaces instances

### PR DESCRIPTION
**Pull request by NaderAlAwar**

### CODE REVIEW

File: examples/pykokkos/streams.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    elif x == 2:
+        pk.printf("Stream 2\n")
+
+if __name__ == "__main__":
+    space = pk.Cuda
+
+    s1 = cp.cuda.Stream()
+    s2 = cp.cuda.Stream()
+
+    instance1 = pk.ExecutionSpaceInstance(space, s1)
```
Comment (gliga) [inline examples/pykokkos/streams.py:20]: ExecutionSpaceInstance => ExecutionSpace

File: examples/pykokkos/streams.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        pk.printf("Default stream\n")
+    elif x == 1:
+        pk.printf("Stream 1\n")
+    elif x == 2:
+        pk.printf("Stream 2\n")
+
+if __name__ == "__main__":
+    space = pk.Cuda
+
+    s1 = cp.cuda.Stream()
```
Comment (gliga) [inline examples/pykokkos/streams.py:17]: Should we use pk.Stream to avoid importing cupy

File: examples/pykokkos/streams.py (hunk 1 above)
Comment (gliga) [reply examples/pykokkos/streams.py:20]: if we are instantiating a stream on top of a space, we could also go with `space.make(stream)`. (And potentially have also a version that does not accept a stream but uses a new stream for each call.)

File: pykokkos/core/translators/bindings.py [hunk 3]
```diff
@@ -311,7 +312,7 @@ def generate_call(operation: str, functor: str, members: PyKokkosMembers, tag: c
         call += f"}} else {{ {custom_call} }}"
 
     call += generate_copy_back(members)
-    call += generate_fence_call()
+    # call += generate_fence_call()
```
Comment (gliga) [inline pykokkos/core/translators/bindings.py:315]: including a comment to explain the reason this is commented out would be good

File: pykokkos/interface/execution_space.py [hunk 4]
```diff
@@ -22,4 +22,38 @@ def is_host_execution_space(space: ExecutionSpace) -> bool:
     if space is ExecutionSpace.Default:
         space = km.get_default_space()
 
-    return space in {ExecutionSpace.OpenMP, ExecutionSpace.Pthreads, ExecutionSpace.Serial}
\ No newline at end of file
+    return space in {ExecutionSpace.OpenMP, ExecutionSpace.Threads, ExecutionSpace.Serial}
+
+class ExecutionSpaceInstance:
+    """
+    An instance of the execution space, corresponding to Cuda/HIP
```
Comment (gliga) [inline pykokkos/interface/execution_space.py:29]: If this only corresponds to streams, should we call it that way?

**Review by gliga — COMMENTED** [review]
Looks good.

A few comments for a future PR.
The difference between execution space and execution space instance is confusing: it could be due to names used, so considering switching to Stream for instances.

File: examples/pykokkos/streams.py (hunk 2 above)
Comment (NaderAlAwar) [reply examples/pykokkos/streams.py:17]: The motivation here is that users might have pre-existing CuPy code they want to integrate with pykokkos, and this example shows how.

File: examples/pykokkos/streams.py (hunk 1 above)
Comment (NaderAlAwar) [reply examples/pykokkos/streams.py:20]: That does seem like a better interface. I'll address this in a new PR

File: pykokkos/interface/execution_space.py (hunk 4 above)
Comment (NaderAlAwar) [reply pykokkos/interface/execution_space.py:29]: Technically users can create an instance for OpenMP (which doesn't have streams), although I don't think it does anything right now. Future execution spaces might have use for this and they might not use the terminology "Streams".

File: examples/pykokkos/streams.py (hunk 2 above)
Comment (gliga) [reply examples/pykokkos/streams.py:17]: but the example only uses cupy to create streams not to use cupy arrays; would be good to then use cupy for all parts

File: pykokkos/interface/execution_space.py (hunk 4 above)
Comment (gliga) [reply pykokkos/interface/execution_space.py:29]: Stream would at least reflect the current usage.  In the future, you could reconsider the name if needed.  `ExecutionSpaceInstace` sounds like an instance of a class, but then instance of an instance sounds like a meta level. This is something to be discussed in the next pykokkos meeting, but seems unnecessarily complex in my view.

**Pull request merged**